### PR TITLE
[Enhancement] Support line delimiter '\r' in hive textfile format

### PR DESCRIPTION
**Pull request by Smith-Cruise**
If the user didn't specify the Hive textfile's `field.delim` in the table's SERDEPROPERTIES, its textfile might use `\r`, `\n`, or `\r\n` as the line delimiter. So our BE needs to handle this situation adaptively.

After this PR, we still don't support `\r\n` as a line delimiter, because it needs to refactor the code.

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:

- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:disappointed: **fail** : 3 / 8 (37.50%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/connector/hive/TextFileFormatDesc.java|3|8|37.50%|[58, 59, 60, 61, 62]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:heart_eyes: **pass** : 41 / 44 (93.18%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/exec/hdfs_scanner_text.cpp|38|41|92.68%|[198, 224, 233]|
|:large_blue_circle:|src/formats/csv/csv_reader.cpp|1|1|100.00%|[]|
|:large_blue_circle:|src/exec/csv_scanner.cpp|2|2|100.00%|[]|

**Review by letian-jiang — APPROVED** [review]

File: be/src/exec/hdfs_scanner_text.cpp [hunk 1]
```diff
@@ -144,27 +156,82 @@ void HdfsScannerCSVReader::_trim_row_delimeter(Record* record) {
...
+    if (LIKELY(!_need_probe_line_delimiter)) {
+        return buffer.find(_parse_options.row_delimiter, pos);
+    } else {
+        // If we didn't get explicit line.delim from hms, we need to probe it.
+        // We will probe '\n' first, most of TextFile's line.delim is '\n'
+        // Then we will try to probe '\r'
+        // TODO(Smith)
+        // We didn't support to treat '\r\n' as line.delim,
+        // because our code does not support line separator's length larger than one char.
+        char* p = buffer.find(DEFAULT_LINE_DELIM, pos);
```
Comment (dirtysalt) [inline be/src/exec/hdfs_scanner_text.cpp:170]: Will we consume a \r if there is \r\n as delimiter?

File: be/src/exec/hdfs_scanner_text.cpp [hunk 2]
```diff
@@ -25,6 +25,13 @@
 
 namespace starrocks {
 
+const std::string DEFAULT_FIELD_DELIM = "\001";
+const std::string DEFAULT_COLLECTION_DELIM = "\002";
+const std::string DEFAULT_MAPKEY_DELIM = "\003";
+// TODO not support \r\n line delimiter yet
+const std::string DEFAULT_LINE_DELIM = "\n";
+const std::string ANOTHER_DEFAULT_LINE_DELIM = "\r";
```
Comment (dirtysalt) [inline be/src/exec/hdfs_scanner_text.cpp:33]: better name like carriage return line delim.

File: be/src/exec/hdfs_scanner_text.cpp (hunk 1 above)
Comment (Smith-Cruise) [reply be/src/exec/hdfs_scanner_text.cpp]: We only support single char delimiter now.

File: be/src/exec/hdfs_scanner_text.cpp (hunk 2 above)
Comment (Smith-Cruise) [reply be/src/exec/hdfs_scanner_text.cpp]: renamed

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:disappointed: **fail** : 3 / 8 (37.50%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/connector/hive/TextFileFormatDesc.java|3|8|37.50%|[58, 59, 60, 61, 62]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:heart_eyes: **pass** : 41 / 44 (93.18%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/exec/hdfs_scanner_text.cpp|38|41|92.68%|[203, 229, 238]|
|:large_blue_circle:|src/formats/csv/csv_reader.cpp|1|1|100.00%|[]|
|:large_blue_circle:|src/exec/csv_scanner.cpp|2|2|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:heart_eyes: **pass** : 12 / 12 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/connector/hive/HiveMetastoreApiConverter.java|4|4|100.00%|[]|
|:large_blue_circle:|com/starrocks/connector/hive/TextFileFormatDesc.java|8|8|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]

:heart_eyes: **pass** : 41 / 44 (93.18%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|src/exec/hdfs_scanner_text.cpp|38|41|92.68%|[203, 229, 238]|
|:large_blue_circle:|src/formats/csv/csv_reader.cpp|1|1|100.00%|[]|
|:large_blue_circle:|src/exec/csv_scanner.cpp|2|2|100.00%|[]|

**Review by dirtysalt — APPROVED** [review]

**Review by Youngwb — APPROVED** [review]

**Pull request merged**